[AST] Store only interface types in NormalProtocolConformances.

DougGregor opened the pull request:
Rather than storing contextual types in the type witnesses and associated
conformances of NormalProtocolConformance, store only interface types.

@huonw did most of the work here, and @DougGregor patched things up to
complete the change.

Part of rdar://problem/34944286.

huonw committed:
[AST] Store only interface types in NormalProtocolConformances.

Rather than storing contextual types in the type witnesses and associated
conformances of NormalProtocolConformance, store only interface types.

@huonw did most of the work here, and @DougGregor patched things up to
complete the change.

DougGregor committed:
Add no-longer-crashing test case from rdar://problem/35441779.

DougGregor commented:
@swift-ci please test and merge

slavapestov commented on lib/AST/ASTMangler.cpp:
```diff
@@ -1978,7 +1978,10 @@ void ASTMangler::appendProtocolConformance(const ProtocolConformance *conformanc
-    auto conformingType = conformance->getType()->mapTypeOutOfContext();
+    auto conformingType = conformance->getType();
+    if (!isa<NormalProtocolConformance>(conformance)) {
+      conformingType = conformingType->mapTypeOutOfContext();
```
This seems bad, just because a conformance is specialized or inherited doesn't mean its written in terms of archetypes. The caller should pass the right conformance in. Can you see if this code path is hit, and if so file a radar?

DougGregor replied:
Yeah, it's weird. I'll see if it can go.

slavapestov commented on lib/IRGen/GenProto.cpp:
```diff
@@ -981,16 +981,17 @@ static bool isDependentConformance(IRGenModule &IGM,
             TypeDecl *explicitDecl) -> bool {
           // RESILIENCE: this could be an opaque conformance
-          return type->hasArchetype();
+          return DC->mapTypeIntoContext(type)->hasArchetype();
```
This can just be `type->hasTypeParameter()`?

DougGregor replied:
I think so!

slavapestov commented on lib/IRGen/GenProto.cpp:
```diff
@@ -1208,7 +1209,7 @@ class AccessorConformanceInfo : public ConformanceInfo {
                         SILWitnessTable *SILWT)
         : IGM(IGM), Table(table),
-          ConcreteType(SILWT->getConformance()->getType()->getCanonicalType()),
+          ConcreteType(SILWT->getConformance()->getDeclContext()->mapTypeIntoContext(SILWT->getConformance()->getType()->getCanonicalType())),
```
This line seems... too long

slavapestov commented on lib/IRGen/GenProto.cpp:
```diff
@@ -2347,7 +2353,13 @@ llvm::Value *MetadataPath::followComponent(IRGenFunction &IGF,
+    if (sourceConformance.isConcrete() &&
+        isa<NormalProtocolConformance>(sourceConformance.getConcrete())) {
+      associatedType =
+        sourceConformance.getConcrete()->getDeclContext()->mapTypeIntoContext(associatedType)
```
Also kidn of lng

slavapestov commented:
This looks great -- just some minor nitpicks

DougGregor committed:
[SIL Printer] Wire up generic environment when printing witness tables.

Now that we have interface types in witness tables, make sure to set the
appropriate generic environment when printing witness tables. This
ensures that we get the resugared generic type parameter names.

DougGregor committed:
Cleanups for interface types in normal conformances.

Fix some line breaks, remove some unnecessary code, and fix a
crash in printing synthesized extensions.

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
@swift-ci please test Linux

DougGregor commented:
@swift-ci please test Linux

DougGregor commented:
@swift-ci please test compiler performance

DougGregor commented:
@swift-ci please test compiler performance
